http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/KubernetesServiceInterp.scala

```diff
@@ -0,0 +1,412 @@
+
+    val defaultNodepool = for {
+      nodepoolName <- KubernetesNameUtils.getUniqueName(NodepoolName.apply)
+    } yield Nodepool(
+      NodepoolLeoId(-1),
+      clusterId = KubernetesClusterLeoId(-1),
+      nodepoolName,
+      status = NodepoolStatus.Precreating,
+      auditInfo,
+      machineType = leoKubernetesConfig.nodepoolConfig.defaultNodepoolConfig.machineType,
+      numNodes = leoKubernetesConfig.nodepoolConfig.defaultNodepoolConfig.numNodes,
+      autoscalingEnabled = leoKubernetesConfig.nodepoolConfig.defaultNodepoolConfig.autoscalingEnabled,
+      autoscalingConfig = None,
+      List.empty,
+      List()
+    )
+
+    for {
+      nodepool <- defaultNodepool
+      clusterName <- KubernetesNameUtils.getUniqueName(KubernetesClusterName.apply)
+    } yield SaveKubernetesCluster(
+      googleProject = googleProject,
+      clusterName = clusterName,
+      location = leoKubernetesConfig.clusterConfig.location,
+      status = KubernetesClusterStatus.Precreating,
+      serviceAccount = leoKubernetesConfig.serviceAccountConfig.leoServiceAccountEmail,
+      auditInfo = auditInfo,
+      defaultNodepool = nodepool
+    )
+  }
+
+  private[service] def getUserNodepool(clusterId: KubernetesClusterLeoId,
+                                       userInfo: UserInfo,
+                                       req: CreateAppRequest,
+                                       now: Instant): Either[Throwable, Nodepool] = {
+    val auditInfo = AuditInfo(userInfo.userEmail, now, None, now)
+
+    val machineConfig = req.kubernetesRuntimeConfig.getOrElse(
+      KubernetesRuntimeConfig(
+        leoKubernetesConfig.nodepoolConfig.galaxyNodepoolConfig.numNodes,
+        leoKubernetesConfig.nodepoolConfig.galaxyNodepoolConfig.machineType,
+        leoKubernetesConfig.nodepoolConfig.galaxyNodepoolConfig.autoscalingEnabled
+      )
+    )
+
+    for {
+      nodepoolName <- KubernetesNameUtils.getUniqueName(NodepoolName.apply)
+    } yield Nodepool(
+      NodepoolLeoId(-1),
+      clusterId = clusterId,
+      nodepoolName,
+      status = NodepoolStatus.Precreating,
+      auditInfo,
+      machineType = machineConfig.machineType,
+      numNodes = machineConfig.numNodes,
+      autoscalingEnabled = machineConfig.autoscalingEnabled,
+      autoscalingConfig = Some(leoKubernetesConfig.nodepoolConfig.galaxyNodepoolConfig.autoscalingConfig),
+      List.empty,
+      List.empty
+    )
+  }
+
+  private[service] def getSavableApp(googleProject: GoogleProject,
+                                     appName: AppName,
+                                     userInfo: UserInfo,
+                                     samResourceId: AppSamResource,
+                                     req: CreateAppRequest,
+                                     diskOpt: Option[PersistentDisk],
+                                     nodepoolId: NodepoolLeoId,
+                                     ctx: AppContext): Either[Throwable, SaveApp] = {
+    val now = ctx.now
```

Comment:
minor, pass in `now: Instant` directly?

Reply:
some of the exceptions print need the traceId from the ctx to print more helpful messages